snapshots/util.go

```diff
@@ -0,0 +1,162 @@
+package snapshots
+
+import (
+	"errors"
+	"io"
+)
+
+// ChunkWriter reads an input stream, splits it into fixed-size chunks, and writes them to a
+// sequence of io.ReadClosers via a channel.
+type ChunkWriter struct {
+	ch        chan<- io.ReadCloser
+	pipe      *io.PipeWriter
+	chunkSize uint64
+	written   uint64
+	closed    bool
+}
+
+// NewChunkWriter creates a new ChunkWriter. If chunkSize is 0, no chunking will be done.
+func NewChunkWriter(ch chan<- io.ReadCloser, chunkSize uint64) *ChunkWriter {
+	return &ChunkWriter{
+		ch:        ch,
+		chunkSize: chunkSize,
+	}
+}
+
+// chunk creates a new chunk.
+func (w *ChunkWriter) chunk() error {
+	if w.pipe != nil {
+		err := w.pipe.Close()
+		if err != nil {
+			return err
+		}
+	}
+	pr, pw := io.Pipe()
+	w.ch <- pr
+	w.pipe = pw
+	w.written = 0
+	return nil
+}
+
+// Close implements io.Closer.
+func (w *ChunkWriter) Close() error {
+	if !w.closed {
+		w.closed = true
+		close(w.ch)
+		var err error
+		if w.pipe != nil {
+			err = w.pipe.Close()
+		}
+		return err
+	}
+	return nil
+}
+
+// CloseWithError closes the writer and sends an error to the reader.
+func (w *ChunkWriter) CloseWithError(err error) {
+	if !w.closed {
+		w.closed = true
+		close(w.ch)
+		if w.pipe != nil {
+			w.pipe.CloseWithError(err)
+		}
+	}
+}
+
+// Write implements io.Writer.
+func (w *ChunkWriter) Write(data []byte) (int, error) {
+	if w.closed {
+		return 0, errors.New("cannot write to closed ChunkWriter")
+	}
+	nTotal := 0
+	for len(data) > 0 {
+		if w.pipe == nil || (w.written >= w.chunkSize && w.chunkSize > 0) {
+			err := w.chunk()
+			if err != nil {
+				return nTotal, err
+			}
+		}
+
+		var writeSize uint64
+		if w.chunkSize == 0 {
+			writeSize = uint64(len(data))
+		} else {
+			writeSize = w.chunkSize - w.written
+		}
+		if writeSize > uint64(len(data)) {
+			writeSize = uint64(len(data))
+		}
+
+		n, err := w.pipe.Write(data[:writeSize])
+		w.written += uint64(n)
+		nTotal += n
+		if err != nil {
+			return nTotal, err
+		}
+		data = data[writeSize:]
+	}
+	return nTotal, nil
+}
+
+// ChunkReader reads chunks from a channel of io.ReadClosers and outputs them as an io.Reader
+type ChunkReader struct {
+	ch     <-chan io.ReadCloser
+	reader io.ReadCloser
+}
+
+// NewChunkReader creates a new ChunkReader.
+func NewChunkReader(ch <-chan io.ReadCloser) *ChunkReader {
+	return &ChunkReader{ch: ch}
+}
```

Comment:
Ideally we wouldn't need to use byte slices at all, and only use IO streams, but unfortunately ABCI does not support streaming. I built it this way so that we could implement a streaming API at some later point (this is quickly becoming a necessity for e.g. `InitChain`, where we currently on chain upgrades have to load the entire state into memory, which isn't going to work for much longer).

Another advantage is that `io.ReadCloser` allows us to propagate IO errors, via `pipe.CloseWithError()`. If we pass byte slices we would have to wrap them in another result struct for error propagation as well.

Reply:
OK, maybe we can add a comment line explaining that.

Reply:
Yup, done